pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,652 @@
+		if checkErr != nil {
+			return false, checkErr
+		}
+		if !ok {
+			return false, nil
+		}
+		ok, checkErr = checkExternalEntityExists(controller.crdClient, *existingEntity)
+		if checkErr != nil {
+			return false, checkErr
+		}
+		if ok {
+			return false, nil
+		}
+		return true, nil
+	})
+	assert.NoError(t, err)
+}
+
+func testAddExternalNode(t *testing.T, externalNode *v1alpha1.ExternalNode, expectedEntities []v1alpha2.ExternalEntity) {
+	controller := newExternalNodeController([]runtime.Object{externalNode})
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	go controller.Run(stopCh)
+	err := wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+		for _, ee := range expectedEntities {
+			ok, err := checkExternalEntityExists(controller.crdClient, ee)
+			if err != nil {
+				return false, err
+			}
+			if !ok {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	assert.NoError(t, err)
+}
+
+func testUpdateExternalNode(t *testing.T, externalNode, updatedExternalNode *v1alpha1.ExternalNode, existingEntity, expectedEntity v1alpha2.ExternalEntity, existsBeforeUpdate bool) {
+	controller := newExternalNodeController([]runtime.Object{externalNode})
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	go controller.Run(stopCh)
+	err := wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+		return checkExternalEntityExists(controller.crdClient, existingEntity)
+	})
+	require.NoError(t, err)
+	if !existsBeforeUpdate {
+		err = controller.crdClient.CrdV1alpha2().ExternalEntities(existingEntity.Namespace).Delete(context.TODO(), existingEntity.Name, metav1.DeleteOptions{})
+		require.NoError(t, err)
+	}
+	_, err = controller.crdClient.CrdV1alpha1().ExternalNodes(externalNode.Namespace).Update(context.TODO(), updatedExternalNode, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	err = wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+		return checkExternalEntityExists(controller.crdClient, expectedEntity)
+	})
+	assert.NoError(t, err)
+	exists, err := checkExternalEntityExists(controller.crdClient, existingEntity)
+	assert.NoError(t, err)
+	assert.False(t, exists)
+}
+
+func testDeleteExternalNode(t *testing.T, externalNode *v1alpha1.ExternalNode, expectedEntities []v1alpha2.ExternalEntity) {
+	controller := newExternalNodeController([]runtime.Object{externalNode})
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	go controller.Run(stopCh)
+	err := wait.PollImmediate(time.Millisecond*50, time.Second, func() (done bool, err error) {
+		for _, ee := range expectedEntities {
+			ok, err := checkExternalEntityExists(controller.crdClient, ee)
+			if err != nil {
+				return false, err
+			}
+			if !ok {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
```

Comment:
I feel this is duplicate of TestAddExternalNode, maybe it should just initialize the storage with the externalEntity and just check deleting externalNode could delete externalEntities eventually.

Reply:
updated.